mantidproject/mantid
# Implement uniform distribution of events in N sphere in FakeMD

**Description of work.**

Previously the orientation/direction of a unit vector was found from uniformly distributed random coords in a hypercube, which were then used to generate an event inside a hypershpere. However this gives greater probability of events near the vertices leading to lines of intensity pointing towards the vertices. Now the orientation is determined from uniform sampling of surface of unit N-sphere (algorithm adapted from [1]).


![image](https://user-images.githubusercontent.com/55979119/156017339-26f869d5-0b7c-42a3-9b6b-3609d6def362.png)


[1] http://corysimon.github.io/articles/uniformdistn-on-sphere/

**To test:**

(1) Run this script
```
ws = CreateMDWorkspace(Dimensions='3', Extents='-5,5,-4,4,-3,3',
                       Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.',
                       Frames='HKL,HKL,HKL',
                       SplitInto='2', SplitThreshold='50')
expt_info = CreateSampleWorkspace()
ws.addExperimentInfo(expt_info)
SetUB(ws, 1,1,1,90,90,90)
p = CreatePeaksWorkspace(InstrumentWorkspace='ws', NumberOfPeaks=0, OutputWorkspace='peaks')
FakeMDEventData(ws, PeakParams='1e+05,0,1,1,0.3', RandomSeed='3873875')
pk = p.createPeakHKL([0,1,1]) # axes aligned with viewing axes
p.addPeak(pk)
```
(2) View in `ws` in sliceviewer
(3) Overlay `peaks`
(4) Double click on first row in peak viewer table on RHS
(5) Should be region of random uniform events within sphere of radius 0.3

Fixes #33213 





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
